KAFKA-10006: Don't create internal topics when LeaderNotAvailableException

showuon opened the pull request:
1. Separate the exception handler for `LeaderNotAvailableException` and `UnknownTopicOrPartitionException`
2. For `UnknownTopicOrPartitionException`, keep as is (i.e. take the topic as not existed)
3. For `LeaderNotAvailableException`, log with correct message, and keep retries until the old leader or new leader become available, or running out of retries
  3.1. add a set `leaderNotAvailableTopics`, to keep the topics that has `LeaderNotAvailableException`
  3.2. Before trying to add to `topicsToCreate` set, check if the topicName is in the `leaderNotAvailableTopics`, if yes, then don't add (i.e. don't create this topic b/c its leader currently is not available)
  3.3. When running out of retries, throw `StreamsException` with correct message
4. add tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon committed:
KAFKA-10006: do not attempt to create internal topics if got LeaderNotAvailableException

showuon committed:
KAFKA-10006: add tests

showuon committed:
KAFKA-10006: refactor

showuon commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -157,16 +161,16 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
-            if (!topicsNotReady.isEmpty()) {
-                log.info("Topics {} can not be made ready with {} retries left", topicsNotReady, retries);
+            if (isNeedRetry(topicsNotReady)) {
+                log.info("Topics {} can not be made ready with {} retries left", topicsNotReady, remainingRetries);
```
the log should put `remainingRetries` since it's logging `with {} retries left`

ableegoldman replied:
Good catch

showuon commented on clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:
```diff
@@ -330,6 +331,12 @@ synchronized public DescribeTopicsResult describeTopics(Collection<String> topic
+            // try to simulate the leader not available situation when topic name is "LeaderNotAvailableTopic"
+            if (requestedTopic.equals("LeaderNotAvailableTopic")) {
+                KafkaFutureImpl<TopicDescription> future = new KafkaFutureImpl<>();
+                future.completeExceptionally(new LeaderNotAvailableException("The leader of Topic " + requestedTopic + " is not available."));
```
Try to simulate the `LeaderNotAvailableException` in the `MockAdminClient`, if the topic name is`LeaderNotAvailableTopic`.

ableegoldman replied:
Is it possible to use `EasyMock` instead of adding this to the actual `MockAdminClient`? I know it's kind of a pain to set up but I think it'll make the test a lot more clear. I did something similar in StreamsPartitionAssignorTest to mock the results of the `listOffsets` request

showuon commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -247,11 +261,19 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
     }
+
+    private boolean isNeedRetry(final Set<String> topicsNotReady) {
+        return !topicsNotReady.isEmpty() || hasNeedRetryTopic();
```
If there's topic with `LeaderNotAvailableException`, we also need to retry.

showuon commented:
Hi @ableegoldman , could you review this PR? Thanks.

showuon committed:
KAFKA-10006: refactor

showuon commented:
Hi @abbccdda @ableegoldman , please help review this PR. Thanks

ableegoldman commented:
Hey @showuon, thanks for the PR! I'm pretty busy at the moment, especially with the 2.6 deadline, but it's on my todo list to review.

showuon commented:
Thanks, @ableegoldman . Take your time. :)

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -59,6 +59,9 @@ private InternalAdminClientConfig(final Map<?, ?> props) {
     private final long retryBackOffMs;
+    private int remainingRetries;
+
+    private HashSet<String> leaderNotAvailableTopics = new HashSet<>();
```
This should be declared final.

abbccdda replied:
Similar to this struct, it doesn't make sense to have a non-empty leaderNotAvailableTopics after each call to `makeReady`, I would prefer building it as local variable, cc @ableegoldman

showuon replied:
Oh, Nice catch! You're correct, @abbccdda , it might have a non-empty leaderNotAvailableTopics after each call to `makeReady`. I changed it to using local variable. Thank you.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -195,20 +198,30 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
-                    topicFuture.getKey(),
-                    topicDescription.partitions().size());
+                existedTopicPartition.put(topicName, topicDescription.partitions().size());
+                if (leaderNotAvailableTopics.contains(topicName)) {
```
This contains check is unnecessary.

showuon replied:
Good point! The `remove` method will return true/false for key found/not found. Fixed.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -100,11 +103,11 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
         log.debug("Starting to validate internal topics {} in partition assignor.", topics);
 
-        int remainingRetries = retries;
+        remainingRetries = retries;
```
This is a personal preference, but I think we should not attempt to include a temporal variable as part of the class struct. We could change the internal function signatures to pass around remainingRetries (like `validateTopics`) instead.

showuon replied:
OK, fixed. Thanks.

abbccdda commented on clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java:
```diff
@@ -330,6 +331,12 @@ synchronized public DescribeTopicsResult describeTopics(Collection<String> topic
                 future.completeExceptionally(new UnknownTopicOrPartitionException("Topic " + requestedTopic + " not found."));
                 topicDescriptions.put(requestedTopic, future);
             }
+            // try to simulate the leader not available situation when topic name is "LeaderNotAvailableTopic"
```
This workaround is very hard to be found by other developers, as a minimum we should define a constant and make it part of `MockAdminClient` class

showuon replied:
Good suggestion. I put it as a constant. Thanks.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -287,12 +290,49 @@ public void shouldLogWhenTopicNotFoundAndNotThrowException() {
+                StreamsException.class,
+                () -> internalTopicManager.makeReady(topicConfigMap));
+
+            final String expectedMessage = "Could not create topics after 1 retries. This can happen if the Kafka cluster is temporary not available";
```
Testing against log message is error-prone and hard to maintain, I think just making sure the thrown exception type is expected should be sufficient.

showuon replied:
Good suggestion. Remove the log message check and only verify the exception thrown. Thanks.

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-10006

showuon committed:
KAFKA-10006: address reviewer's comments

showuon commented:
Thanks @abbccdda , I've address your review comments in this commit: https://github.com/apache/kafka/pull/8712/commits/0567ecbc82c01b0c6e596fe5c28918f3b530a263. Thank you.

showuon commented:
hi @abbccdda @ableegoldman , sorry, but this PR stays for a while, do you have other comments? Or should I find someone else to review? Thanks.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -98,9 +98,10 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
         int remainingRetries = retries;
         Set<String> topicsNotReady = new HashSet<>(topics.keySet());
         final Set<String> newlyCreatedTopics = new HashSet<>();
+        final HashSet<String> leaderNotAvailableTopics = new HashSet<>();
```
Can we give this a more descriptive name? It might be obvious to you, but I think someone just looking at this code for the first time would not get that this actually means topics that may or may not already exist.
That said, I'm struggling to think of a good alternative...maybe `possiblyCreatedTopics` or `unknownTopics`...any better ideas?

showuon replied:
Good suggestion, I changed to `tempUnknownTopics`, because they are temporarily unknown topics for now. How do you think?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -242,11 +256,16 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
+
+    private boolean shouldRetry(final Set<String> topicsNotReady, final HashSet<String> leaderNotAvailableTopics) {
+        // If there's topic with LeaderNotAvailableException, we still need retry
+        return !topicsNotReady.isEmpty() || leaderNotAvailableTopics.size() > 0;
```
Can we just use `!isEmpty` for both sets?

showuon replied:
Good suggestion. Updated.

ableegoldman reviewed (commented):
Hey @showuon , thanks for the PR! And especially thanks for being so patient. Sorry for the long wait.

Left a few comments but this looks good in general. My main question is whether we can rework the test a bit to avoid hacking this specific case into the MockAdminClient?

showuon committed:
KAFKA-10006: use EasyMock to rewrite the test, and remove the unneeded hack in MockAdminClient

showuon commented:
Hi @ableegoldman , thanks for the good suggestion! I've updated in this commit: https://github.com/apache/kafka/pull/8712/commits/f378c34d10c57d796412ac946f867e5787f93af0. What I did in this commit is:

1. rewrite the test by `EasyMock`, to make the tests better, and because of this test rewrite, I found a bug that I missed before, which is that when we retried the `describeTopics` in `getNumPartitions`, we should also add the `tempUnknownTopics` into the to-be-describe list since they also needed to be retried. Thanks for the suggestion! 
2. remove unneeded hack in MockAdminClient
3. rename the variable name `leaderNotAvailableTopics` to `tempUnknownTopics`
4. use `isEmpty()` to be consistent

Thank you.

abbccdda commented:
retest this

ableegoldman reviewed (approved):
Thanks for the fix! This LGTM

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -98,9 +98,10 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
         int remainingRetries = retries;
         Set<String> topicsNotReady = new HashSet<>(topics.keySet());
         final Set<String> newlyCreatedTopics = new HashSet<>();
+        final HashSet<String> tempUnknownTopics = new HashSet<>();
```
s/HashSet/Set?

showuon replied:
Fixed. Thanks.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -243,10 +259,18 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
                 }
             } else {
-                topicsToCreate.add(topicName);
+                // for the tempUnknownTopics, we'll check again later if retries > 0
```
Could be merged with above `else`

showuon replied:
Fixed

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -287,12 +291,41 @@ public void shouldLogWhenTopicNotFoundAndNotThrowException() {
+
+        // simulate describeTopics got LeaderNotAvailableException
+        EasyMock.expect(admin.describeTopics(Collections.singleton(leaderNotAvailableTopic)))
+                .andReturn(new MockDescribeTopicsResult(
```
Use 4 space format to align with other tests.

showuon replied:
Nice catch! Fixed.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -179,31 +180,43 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
-        final DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(topics);
+    protected Map<String, Integer> getNumPartitions(final Set<String> topics,
+                                                    final HashSet<String> tempUnknownTopics,
+                                                    final int remainingRetries) {
```
We could just pass in a boolean here to indicate whether there are remaining retries

showuon replied:
Updated.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -179,31 +180,43 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
+                        "Error message was: {}", topicName, cause.toString());
+                } else if (cause instanceof LeaderNotAvailableException) {
+                    tempUnknownTopics.add(topicName);
+                    if (remainingRetries > 0) {
```
Could reduce the if-else block as:
```
if (remainingRetries <= 0) {
  // run out of retries, throw exception directly 
  throw new StreamsException(
    String.format("The leader of the Topic %s is not available after %d retries.", topicName, retries), cause);
 }
 log.debug("The leader of the Topic {} is not available, with {} retries left.\n" +
   "Error message was: {}", topicName, remainingRetries, cause.toString());

```

showuon replied:
Updated.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -179,31 +180,43 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
+                                                    final HashSet<String> tempUnknownTopics,
+                                                    final int remainingRetries) {
+        final Set<String> allTopicsToDescribe = new HashSet<>(topics);
+        allTopicsToDescribe.addAll(tempUnknownTopics);
```
Why do we need `allTopicsToDescribe`? It seems only queried once locally.

showuon replied:
The `allTopicsToDescribe` is for `tempUnknownTopics` to have chance to get described again, by the retries in `makeReady` method. In the `makeReady`, we want to know which topics existed and to validate it, and which topics not existed that needed to be created. But for the `LeaderNotAvailable`topics, we can't know if topics existed or not. So, we need to merge `topics`(topics to validate) and `tempUnknownTopics` here to describe them all (again) here.

Thanks.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -218,13 +231,16 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
-    private Set<String> validateTopics(final Set<String> topicsToValidate, final Map<String, InternalTopicConfig> topicsMap) {
+    private Set<String> validateTopics(final Set<String> topicsToValidate,
+                                       final Map<String, InternalTopicConfig> topicsMap,
+                                       final HashSet<String> tempUnknownTopics,
```
Similar here, we could reduce to Set

showuon replied:
Fixed.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -108,7 +111,8 @@ public void shouldReturnCorrectPartitionCounts() {
             null);
-        assertEquals(Collections.singletonMap(topic, 1), internalTopicManager.getNumPartitions(Collections.singleton(topic)));
+        assertEquals(Collections.singletonMap(topic, 1),
+                internalTopicManager.getNumPartitions(Collections.singleton(topic), new HashSet<String>(), 1));
```
Could use Collections.emptySet() if reduced to Set

showuon replied:
Updated. Thanks.

abbccdda reviewed (commented):
Thanks for the PR, get a couple of minor comments.

showuon committed:
KAFKA-10006: address reviewer's comments

showuon commented:
hi @abbccdda , thanks for the comments. I've updated them in this commit: https://github.com/apache/kafka/pull/8712/commits/70d41212fe8a4d5055993c56e30c2342498e1664. Please help review again. Thanks.

showuon commented:
hi @abbccdda , Last night before I went to sleep, I kept thinking about your question: 
> Why do we need allTopicsToDescribe?

Then I found my change has 1 small logic flaw, that is if the topic after `LeaderNotAvailableException` thrown, it got `UnknownTopicException`, we won't create topic for it. I'll fix it, and then add 2 more tests for 2 different situation. Will let you know. Thanks.

showuon committed:
KAFKA-10006: fix the log flaw to count the tempUnkonwnTopics as topicsNotReady and add 2 more tests

1. Return the topicsNotReady to makeReady including tempUnknownTopics, and not create topic to wait for next retry
2. tempUnknownTopics will be created each retry since we count the tempUnknownTopics as part of topicsNotReady
3. add 2 more tests to total test 3 cases:
  3.1 shouldCreateTopicWhenTopicLeaderNotAvailableAndThenTopicNotFound
  3.2 shouldCompleteValidateWhenTopicLeaderNotAvailableAndThenDescribeSuccess
  3.3 shouldThrowExceptionWhenKeepsTopicLeaderNotAvailable

showuon commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -125,35 +131,37 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
-                                "Error message was: {}", topicName, cause.toString());
-                            throw new StreamsException(String.format("Could not create topic %s.", topicName), cause);
+                if (createTopicsResult != null) {
+                    for (final Map.Entry<String, KafkaFuture<Void>> createTopicResult : createTopicsResult.values().entrySet()) {
```
Need the null check for `createTopicsResult` since the `newTopics` might be empty

showuon commented:
hi @abbccdda , I've fixed the issue. What I did in the final commit are:
   0. Revert the unnecessary changes 
    1. Return the `topicsNotReady` to `makeReady` including `tempUnknownTopics`, and not create topic to wait for next retry
    2. Empty `tempUnknownTopics` will be created in each retry The goal of `tempUnknownTopics` is to filter out the topics that don't need to create topic in `topicsNotReady` (please check below diagram)
    3. add 2 more tests to total test 3 cases:
      3.1 shouldCreateTopicWhenTopicLeaderNotAvailableAndThenTopicNotFound
      3.2 shouldCompleteValidateWhenTopicLeaderNotAvailableAndThenDescribeSuccess
      3.3 shouldThrowExceptionWhenKeepsTopicLeaderNotAvailable

Below is the sequence diagram for `makeReady`. Hope it's clear for what I did. Thanks.
<img width="1367" alt="image" src="https://user-images.githubusercontent.com/43372967/85853920-ee33fe00-b7e5-11ea-8710-6bdc21bf44e7.png">

abbccdda commented:
retest this please

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -100,13 +100,19 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
 
         while (!topicsNotReady.isEmpty() && remainingRetries >= 0) {
-            topicsNotReady = validateTopics(topicsNotReady, topics);
+            final Set<String> tempUnknownTopics = new HashSet<>();
```
Better to be name as `leaderUnavailableTopics` to match the error code.

showuon replied:
Well, this variable naming is actually suggested by @ableegoldman , and I also think the `tempUnknownTopics` is more descriptive. Is that OK for you? 

https://github.com/apache/kafka/pull/8712#discussion_r444555724

![image](https://user-images.githubusercontent.com/43372967/86199833-b6cea400-bb8d-11ea-81b5-7ba5b6064709.png)

abbccdda replied:
Sure, I don't feel strong about reverting the naming once we already got a reason there.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -287,12 +290,124 @@ public void shouldLogWhenTopicNotFoundAndNotThrowException() {
+        internalTopicConfig.setNumberOfPartitions(1);
+        final Map<String, InternalTopicConfig> topicConfigMap = new HashMap<>();
+        topicConfigMap.put(leaderNotAvailableTopic, internalTopicConfig);
+        topicManager.makeReady(topicConfigMap);
```
Could be simplified as `topicManager.makeReady(Collections.singletonMap(leaderNotAvailableTopic, internalTopicConfig));`

showuon replied:
good. Thanks.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -287,12 +290,124 @@ public void shouldLogWhenTopicNotFoundAndNotThrowException() {
+    }
+
+    @Test
+    public void shouldCompleteValidateWhenTopicLeaderNotAvailableAndThenDescribeSuccess() {
```
This test seems to be overlapping with `shouldCreateTopicWhenTopicLeaderNotAvailableAndThenTopicNotFound`. I don't think we need both to return `LeaderNotAvailable` unless they are evaluating different scenarios.

showuon replied:
Yes, they are different scenarios. You can check below diagram for reference. (red and green for different cases)

![image](https://user-images.githubusercontent.com/43372967/86200914-905e3800-bb90-11ea-8d2c-c5e9c7b71cda.png)

abbccdda replied:
Thanks for explaining! May I know what this software you uses to generate the graph?

showuon replied:
Actually, I drew it by myself on the MS office powerpoint. haha

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -287,12 +290,124 @@ public void shouldLogWhenTopicNotFoundAndNotThrowException() {
+        EasyMock.expect(admin.createTopics(Collections.emptySet()))
+            .andReturn(new MockCreateTopicsResult(Collections.emptyMap())).once();
+
+        EasyMock.expect(admin.describeTopics(Collections.singleton(leaderNotAvailableTopic)))
```
nit: if it will return `topicDescriptionSuccessFuture`, then we should not use `leaderNotAvailableTopic`

showuon replied:
good. Thanks.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -287,12 +290,124 @@ public void shouldLogWhenTopicNotFoundAndNotThrowException() {
+        final Map<String, InternalTopicConfig> topicConfigMap = new HashMap<>();
+        topicConfigMap.put(leaderNotAvailableTopic, internalTopicConfig);
+
+        final StreamsException exception = assertThrows(
```
exception not used.

showuon replied:
Nice catch!

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -287,12 +290,124 @@ public void shouldLogWhenTopicNotFoundAndNotThrowException() {
+        EasyMock.expect(admin.describeTopics(Collections.singleton(leaderNotAvailableTopic)))
+            .andReturn(new MockDescribeTopicsResult(
+                Collections.singletonMap(leaderNotAvailableTopic, topicDescriptionFailFuture)))
+            .times(2);
```
nit: we could set a final int for numRetries as:
```
            put(StreamsConfig.adminClientPrefix(StreamsConfig.RETRIES_CONFIG), numRetries);

```
and use (numRetries + 1) here to clearly indicate we are trying to go beyond the retry limit.

showuon replied:
good suggestion!

abbccdda reviewed (commented):
Thanks for the updates, a few more comments.

showuon committed:
KAFKA-10006: address reviewer's comments

1. simplifiy code
2. use the existing topic name in tests, instead of creating a new one
3. use numRetries to indicate we are trying to go beyond the retry limit

showuon commented:
Thanks for the comments, @abbccdda . I've address your comments in this commit: https://github.com/apache/kafka/pull/8712/commits/4b57a606a3834323302b6d3d33ab95e5b88d183b

What I did in this commit are:
1. simplifiy code
2. use the existing topic name in tests, instead of creating a new topic name: `leaderNotAvailableTopic`
3. use `numRetries` to indicate we are trying to go beyond the retry limit

abbccdda reviewed (commented):
Thanks for the updates, just one more minor comment @showuon

showuon committed:
KAFKA-10006: remove duplicate properties

showuon commented:
@abbccdda , thanks for pointing out the duplicate line. I didn't notice it, My bad! I've fixed. Thanks.

abbccdda commented:
retest this please

abbccdda reviewed (approved):
LGTM, thanks for the patch!

abbccdda commented:
Seems only flaky tests, rerun the test one more time to confirm.

abbccdda commented:
retest this please

abbccdda commented:
Tracked in https://issues.apache.org/jira/browse/KAFKA-9831

abbccdda commented:
Still only flaky tests are failing, merging.
